### PR TITLE
Feature/use ormlite

### PR DESCRIPTION
**Pull request by shervinw**
4 tests fail at -- test/features/cancellation/customer-requested-cancellation/implemented/book-and-cancel-test.js.

Seems to be due to a descrepency between openactive-test-suite and OpenActive.Server.NET. Whereby the test suite is not accepting an empty body on OrderItem cancellation.

According to spec an empty body is required. [https://openactive.io/open-booking-api/EditorsDraft/#customer-requested-cancellation](https://openactive.io/open-booking-api/EditorsDraft/#customer-requested-cancellation).

Submitting this PR on the assumption that it's the test suite that needs modifying.

### CODE REVIEW

**Comment by nickevansuk** [comment]
Great, thanks, will take a look! Could you make one quick update BaseUrl to include https? That was missed in the revert which is why the CI output has issues

**Comment by nickevansuk** [comment]
Great thanks, looks like CI still failing... might be that default.json hasn’t been updated with the new Seller “1” and “2”?

Will take a proper look in the morning if not!

**Comment by shervinw** [comment]
Ahh yes that is the reason!

File: Examples/BookingSystem.AspNetCore/Feeds/OrdersFeed.cs [hunk 1]
```diff
@@ -15,63 +16,74 @@ public class AcmeOrdersFeedRPDEGenerator : OrdersRPDEFeedModifiedTimestampAndID
...
-                            }
-                        };
+                            }).ToList()
+                        }
+                    });
 
-            // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
-            var items = query.ToList().Take(this.RPDEPageSize).ToList();
+                // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
+                var items = query.ToList().Take(this.RPDEPageSize).ToList();
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Feeds/OrdersFeed.cs:74]: We should be able to remove this hack now, as hopefully not an issue with SQLite?  Should we be using `query.Take()` direct to ensure only the `RPDEPageSize` number of rows is returned?

File: Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs [hunk 2]
```diff
@@ -76,9 +77,17 @@ protected override void GetOrderItems(List<OrderItemContext<SessionOpportunity>>
 
             // Response OrderItems must be updated into supplied orderItemContexts (including duplicates for multi-party booking)
 
+            List<OccurrenceTable> occurrenceTable;
+            List<ClassTable> classTable;
+            using (var db = FakeBookingSystem.Database.Mem.Database.Open())
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs:82]: This is interesting... doesn't `using` close the connection? How is this still accessible afterwards?

File: Examples/BookingSystem.AspNetFramework/Feeds/OrdersFeed.cs [hunk 3]
```diff
@@ -15,63 +16,74 @@ public class AcmeOrdersFeedRPDEGenerator : OrdersRPDEFeedModifiedTimestampAndID
...
-                            }
-                        };
+                            }).ToList()
+                        }
+                    });
 
-            // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
-            var items = query.ToList().Take(this.RPDEPageSize).ToList();
+                // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
+                var items = query.ToList().Take(this.RPDEPageSize).ToList();
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetFramework/Feeds/OrdersFeed.cs:74]: Remove hack (as above)?

File: Fakes/OpenActive.FakeDatabase.NET/DatabaseTables.cs [hunk 4]
```diff
@@ -82,58 +97,20 @@ public class SellerTable : Table
...
-        public static string ConvertColumnType(Type type)
-        {
-            if (TypeLookup.ContainsKey(type))
-            {
-                return TypeLookup[type];
-            } else
+            using (var db = dbFactory.Open())
             {
-                return "TEXT NULL";
+                db.DropTable<OrderItemsTable>();
```
Comment (nickevansuk) [inline Fakes/OpenActive.FakeDatabase.NET/DatabaseTables.cs:104]: This is so much cleaner!! Love it!

File: Examples/BookingSystem.AspNetFramework/Feeds/SessionsFeeds.cs [hunk 5]
```diff
@@ -15,161 +16,167 @@ public class AcmeScheduledSessionRPDEGenerator : RPDEFeedModifiedTimestampAndIDL
...
+                        Duration = x.End - x.Start,
+                        RemainingAttendeeCapacity = x.RemainingSpaces,
+                        MaximumAttendeeCapacity = x.TotalSpaces
+                    }
+                });
 
-            // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
-            var items = query.ToList().Take(this.RPDEPageSize).ToList();
+                // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
+                var items = query.ToList().Take(this.RPDEPageSize).ToList();
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetFramework/Feeds/SessionsFeeds.cs:59]: As above, could remove hack?

File: Examples/BookingSystem.AspNetCore/Feeds/SessionsFeeds.cs [hunk 6]
```diff
@@ -15,161 +16,166 @@ public class AcmeScheduledSessionRPDEGenerator : RPDEFeedModifiedTimestampAndIDL
...
+                        Duration = x.End - x.Start,
+                        RemainingAttendeeCapacity = x.RemainingSpaces,
+                        MaximumAttendeeCapacity = x.TotalSpaces
+                    }
+                });
 
-            // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
-            var items = query.ToList().Take(this.RPDEPageSize).ToList();
+                // Note there's a race condition in the in-memory database that allows records to be returned from the above query out of order when modified at the same time. The below ensures the correct order is returned.
+                var items = query.ToList().Take(this.RPDEPageSize).ToList();
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Feeds/SessionsFeeds.cs:59]: As above, remove hack?

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs [hunk 7]
```diff
@@ -43,25 +45,23 @@ public InMemorySQLite()
             // in-memory database
             //const string ConnectionString = "Data Source=/Users/nick/repos/openactive/openactive-test-suite/testsuite.db;Version=3;";
             //const string ConnectionString = "Data Source=InMemoryDatabase;Mode=Memory;Cache=Shared;Version=3;";
-            const string ConnectionString = "Data Source=:memory:?cache=shared";
+            const string ConnectionString = ":memory:";
```
Comment (nickevansuk) [inline Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:48]: Seems much simpler! How come this doesn't need `?cache=shared"`?

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs [hunk 8]
```diff
@@ -43,25 +45,23 @@ public InMemorySQLite()
...
 
             // The in-memory database only persists while a connection is open to it. To manage
             // its lifetime, keep one open master connection around for as long as needed.
             // In this case, for the lifetime of the application.
             PersistentConnection = SQLiteFactory.Instance.CreateConnection();
             PersistentConnection.ConnectionString = ConnectionString;
-            PersistentConnection.Open();
+            // PersistentConnection.Open();
 
             // By default NPoco will open and close a database connection around each query
```
Comment (nickevansuk) [inline Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:58]: Should this comment still reference NPoco? Is this the case for ormite also?

File: Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs (hunk 2 above)
Comment (shervinw) [reply Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs:82]: the references assigned are outside the using statement, this was done for simplicity, so retaining the orginal query syntax.

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs (hunk 7 above)
Comment (shervinw) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:48]: OrmLite automatically registers a `MemoryCacheClient()` if no caching provider is registered explicitly.
ref: [](https://docs.servicestack.net/caching)

Comment (nickevansuk) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:48]: Interesting... I'd understood previously that without `?cache=shared` each new connection would access a completely new database. Perhaps there's some other magic going on here... no worries!

File: Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs (hunk 2 above)
Comment (nickevansuk) [reply Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs:82]: Ok great!

**Comment by nickevansuk** [comment]
Noting we identified that this PR exhibits an intermittent error, where B fails resulting in this response:

```json
{
  "@context": "https://openactive.io/",
  "@type": "OrderCreationFailedError",
  "name": "B failed for a reason not specified by a defined type of error in the Open Booking API specification.",
  "statusCode": 500,
  "description": "Booking failed for an unexpected reason"
}
```

The source of this error is here:

https://github.com/openactive/OpenActive.Server.NET/blob/034c4d7a2ba36f708a91ad3ddecf3445dfd4eb7f/Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs#L249

Because this line is returning `null`:
https://github.com/openactive/OpenActive.Server.NET/blob/034c4d7a2ba36f708a91ad3ddecf3445dfd4eb7f/Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs#L236

Unfortunately `null` is currently returned in two edge cases, which makes it hard to identify the actual source of the error:

https://github.com/openactive/OpenActive.Server.NET/blob/034c4d7a2ba36f708a91ad3ddecf3445dfd4eb7f/Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs#L318-L326

Merging this for now, however if this intermittent error resurfaces we should improve the descriptiveness of the error returned above to identify the root cause.

**Pull request merged**